File: deferred/automation/lib/python/community/deferred.py
"""
Copyright June 25, 2020 Richard Koshak

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
"""
from core.jsr223.scope import events
from org.joda.time import DateTime
from community.time_utils import to_datetime
from community.timer_mgr import TimerMgr

timers = TimerMgr()

def timer_body(target, value, is_command, time, log):
    """
    Called when the differed action timer expires, sends the command to the
    target Item.
    Arguments:
        - target: Item name to send the command to
        - value: Command or state to issue to the target Item
        - is_command: Whether to send value to target as a command or an update
        - time: the original time string
        - log: logger passed in from the Rule that is using this.
    """
    log.debug("{} {} to {} after {}"
              .format("Commanding" if is_command else "Updating", target, value,
                      time))
    if is_command:
        events.sendCommand(target, str(value))
    else:
        events.postUpdate(target, str(value))

def defer(target, value, when, log, is_command=True):
    """
    Use this function to schedule a command to be sent to an Item at the
    specified time or after the speficied delay. If the passed in time or delay
    ends up in the past, the command is sent immediately.
    Arguments:
        - target: Item name to send the command to
        - value: the command to send the Item
        - when: at what time to delay to action until, see to_datetime in the timer_utils library
        - is_command: whether to send value to target as an update or command,
        defaults to True
        - log: logger passed in from the Rule
    """

    trigger_time = to_datetime(when, log)

    if not trigger_time:
        log.error("Cannot schedule a deferred action, {} is not a valid date time or duration"
                  .format(when))

    # If trigger_time is in the past, schedule for now
    if trigger_time.isBefore(DateTime.now()):
        trigger_time = DateTime.now()

    # Schedule the timer
    func = lambda: timer_body(target, value, is_command, when, log)
    flap = lambda: log.debug("there is already a timer set for {}, rescheduling"
                             .format(target))
    timers.check(target, trigger_time, function=func, flapping_function=flap,
                 reschedule=True)

def cancel(target):
    """
    Cancels the timer associated with target if it exists.
    Arguments:
        - target: the Item name whose timer is to be cancelled
    """
    timers.cancel(target)

def cancel_all():
    """
    Cancels all timers.
    """
    timers.cancel_all()

File: deferred/javascript/automation/lib/javascript/community/deferred.js
/**
 * Class that handles deferring a command or update action until a later time.
 */
var Deferred = function() {
    'use strict';
  
    var OPENHAB_CONF = java.lang.System.getenv("OPENHAB_CONF");
    this.ZonedDateTime = Java.type("java.time.ZonedDateTime");
    this.log = Java.type("org.slf4j.LoggerFactory").getLogger("org.openhab.model.script.Rules.Deferred");
    this.log.debug("Building the Deferred object.");
    load(OPENHAB_CONF+'/automation/lib/javascript/community/timerMgr.js');
    this.timers = new TimerMgr();	
    this.log.debug("Deferred is ready to operate");
  }
 
  /**
   * Generates the function that gets called when the timer expires.
   * 
   * @param {string} target The name of the Item
   * @param {string} value The command or update to send to the Item
   * @param {boolean} isCommand when true value will be sent to target as a command, as an update when false
   * @param {Logger} log this class's internal logger
   * @returns function with no arguments that sends a command or update to target 
   */
  Deferred.prototype._timerBodyGenerator = function(target, value, isCommand, time, log) {
    return function() {
      log.debug(((isCommand) ? "Commanding " : "Updating ") + target + " to " + value + " after " + time);
      (isCommand) ? events.sendCommand(target, value) : events.postUpdate(target, value);
    }	  
  }

  /**
   * Generates the function that gets called when a timer is rescheduled.
   * 
   * @param {String} target name of the Item
   * @param {Logger} log this class's internal logger
   */
  Deferred.prototype._rescheduledBodyGenerator = function(target, log){
    return function() {
      log.debug("There is already a timner set for " + target + ", rescheduling");
    }
  }

  /**
   * Waits up to the given amount of time to send a command or update to an Item.
   * 
   * @param {string} target name of the Item
   * @param {string} value state or command to send
   * @param {*} when any representation of time supported by time_utils
   * @param {boolean} isCommand optional flag when true causes the value to be sent as a command
   */
  Deferred.prototype.defer = function(target, value, when, isCommand) {
    var triggerTime = toDateTime(when);
    if(triggerTime.isBefore(this.ZonedDateTime.now())) {
      triggerTime = this.ZonedDateTime.now();
    }
    
    this.timers.check(target, triggerTime, 
                      this._timerBodyGenerator(target, value, isCommand, when, this.log),
                      true,
                      this._rescheduledBodyGenerator(target, this.log));
  } 

  /**
   * Cancel the deferred action for target
   * 
   * @param {*} target name of the Item
   */
  Deferred.prototype.cancel = function(target){
    this.timers.cancel(target);	 
  }
  
  /**
   * Cancels all the deferred actions.
   */
  Deferred.prototype.cancelAll = function() {
    this.timers.cancelAll()
  }